### PR TITLE
feat(DataTable): introduce expand all

### PR DESCRIPTION
**Pull request by asudoh**
Refs #679.

Currently a prototype phase. Created this PR because I feel a need of UX feedback (before the detailed design comes).

The feedback I need is the expando button at the table header, which tends to have "intermediate state". The prototype shows that the expando button as collapsed of any of the rows is collapsed (even though some of the rows are still expanded), demonstrated at https://deploy-preview-2688--carbon-components-react.netlify.com/?selectedKind=DataTable&selectedStory=with%20batch%20expansion. The question is whether it makes sense, or we should come up with the UI for "intermediate state". CC @shixiedesign @laurenmrice 

#### Changelog

**New**

- Styles and JavaScript code to support "expand all" feature in data table.

#### Testing / Reviewing

Testing should make sure data table is not broken.

### CODE REVIEW

**Comment by shixiedesign** [comment]
Thanks for the prototype! I understand why it would be inconsistent with the behavior of checkboxes, but personal I'm feel this prototype works as expected, even without the indeterminate state. If this is the case with y'all, I can create the spec using a chevron like Akira has shown here. @aagonzales @laurenmrice

**Comment by shixiedesign** [comment]
@asudoh please see https://github.com/carbon-design-system/carbon/issues/679#issuecomment-492065365 for design spec! Sorry about the wait. Thank you!

**Comment by aagonzales** [comment]
Yeah I'm not sure what the indeterminate state for this would be, maybe rotate the chevron 90 degrees but not sure how that reads. But I agree with Shixie, I think it works fine without an indeterminate.

File: packages/components/src/components/data-table/_data-table-sort.scss [hunk 1]
```diff
@@ -18,7 +18,7 @@
   // -------------------------------------
   .#{$prefix}--data-table--sort th,
   .#{$prefix}--data-table--sort
-    th:first-of-type:not(.#{$prefix}--table-column-checkbox) {
+    th:first-of-type:not(.#{$prefix}--table-column-checkbox):not(.#{$prefix}--table-expand) {
     padding: 0;
```
Comment (emyarod) [inline packages/components/src/components/data-table/_data-table-sort.scss:22]: it looks like this rule is causing the expand all chevron to be misaligned (but only in Firefox)
image

the expand all chevron needs a top padding of `.875rem` like the other chevrons

**Review by emyarod — APPROVED** [review]
just one alignment issue with the expand all chevron, otherwise looks good to me!

File: packages/components/src/components/data-table/_data-table-sort.scss (hunk 1 above)
Comment (asudoh) [reply packages/components/src/components/data-table/_data-table-sort.scss:22]: Good catch @emyarod - Fixed.

**Review by shixiedesign — COMMENTED** [review]
Hi @asudoh this is great! I only have 1 comment. When all rows are expanded manually, the batch expansion chevron should update to pointing up too.  See below gif. But this is not a stop ship if not possible or requires too much dev time. Thanks you! 

May-20-2019 09-16-58

**Comment by asudoh** [comment]
Good point @shixiedesign - Updated.

**Review by shixiedesign — APPROVED** [review]
Looks good! Thanks! 👍

**Pull request merged**